tests/installation/yast2_firstboot.pm

```diff
@@ -46,45 +47,46 @@ sub license {
     send_key $cmd{next};
 }
 
-sub welcome {
+sub firstboot_welcome {
     assert_screen 'welcome';
     wait_screen_change(sub { send_key $cmd{next}; }, 7);
 }
 
-sub clock_and_timezone {
+sub firstboot_timezone {
     assert_screen 'inst-timezone';
     wait_screen_change(sub { send_key $cmd{next}; }, 7);
 }
 
-sub user_setup {
-    my $is_not_shared_passwd = shift;
+sub firstboot_user {
     assert_screen 'local_user';
     enter_userinfo(username => get_var('YAST2_FIRSTBOOT_USERNAME'));
-    if (defined($is_not_shared_passwd)) {
-        send_key 'alt-t' if (is_opensuse);
-    }
-    else {
-        send_key 'alt-t' if (is_sle('>=12-sp4'));
-    }
     wait_screen_change(sub { send_key $cmd{next}; }, 7);
     await_password_check;
-    wait_screen_change(sub { send_key $cmd{next}; }, 7) unless defined($is_not_shared_passwd);
 }
 
-sub root_setup {
+sub firstboot_root {
     assert_screen 'root_user', 60;
     enter_rootinfo;
     wait_screen_change(sub { send_key $cmd{next}; }, 7);
 }
 
+sub firstboot_hostname {
+    assert_screen 'hostname';
+    wait_screen_change(sub { send_key $cmd{next}; }, 7);
+}
+
 sub run {
-    my $self = shift;
-    language_and_keyboard;
-    welcome;
-    $self->license;
-    clock_and_timezone;
-    user_setup(1);
-    root_setup;
+    my $self      = shift;
+    my $test_data = get_test_suite_data();
+    my %clients;
+    # Make sure the subroutine called from test data exists, by creating a hash ??
+    foreach my $client (@{$test_data->{clients}}) {
+        $clients{$client} = $client;
+    }
+    foreach my $client (keys %clients) {
+        $self->$client->{$client};
```

Comment:
yeah, too many clients here, and it does not even work

Reply:
It seems like keys re-orders the hash somehow, so the functions get called in the wrong order : http://waaa-amazing.suse.cz/tests/12818#step/yast2_firstboot/7

Reply:
sure, that is normal behavior with hashes, only list keep the order.

Reply:
that comes to my mind you have two options to avoid that double loop:
```perl
my %actions = ( language_keyboard => \&language_keyboard,
                  welcome           => \&welcome,
                  licenses          => \&licenses,
                  timezone          => \&timezone,
                  user              => \&user,
                  root              => \&root );
  foreach my $client (@{$test_data->{clients}}) {
          # Option 1 recommended
          #$actions{$client}->();
          #Option 2 really ugly and error-prone
          ( \&$client )->();
  }
```

Reply:
I also recommend you to keep original names of the function, feels tired typing so much :) when prefixing 'firstboot_' to everything in a file that is all about it.

Reply:
Yeah, we cannot rely on the order of the hash keys, as those are ordered by the hash value. So we definitely need a lost there. Nevertheless, we can avoid mapping @jknphy has proposed, as can do reflection, as client module name should correspond to the function name. See example here: https://stackoverflow.com/questions/6286502/call-a-perl-modules-method-from-a-string

Reply:
So here is the code snippet explaining how to do reflection on the module:
```perl
package FooBar;

sub test_method {
    print("Executed");
}

my $method = \&{"FooBar::test_method"};
$method->();
```

With this we could just define modules to call. And in case we need to pass some parameters to those in future, we also could put them in the test_data and pass to the method.

Reply:
The second option that I proposed is exactly that, instead of having the mapping I like your idea of having intermediate variable to use the magic keywords of the language, although with the mapping the loop would looks nicer. Both are the same thing :)

Reply:
Sorry for being such a noob, but I have no idea what you are suggesting now. Also I don't even really see what the problem is with the original loop. I kinda understood the idea of the mapping, even though I didn't really see the point, but now I 'm just completely lost.

Reply:
Also, there is this in the original code "$self->license;"

Reply:
Let me draft the proposal, so it's more clear:
```perl
    # Make sure the subroutine called from test data exists, by creating a hash ??
    foreach my $client (@{$test_data->{clients}}) {
        die "Client '$client' is not defined in the module, please check test_data" unless defined(&{"$client"});
        my $client_method = \&{"$client"};
        $client_method->($self);
    }
```
I'm not 100% sure if that works, but that will allow us to get rid of the second loop.

Reply:
Thanks ! That's starting to make sense now

Reply:
Yeah, sometimes code is more clear than words. Sorry for the confusion ;)

Reply:
Works :) http://waaa-amazing.suse.cz/tests/12825#